src/document-fetcher.js

```diff
@@ -0,0 +1,164 @@
+import {isFormDataWrapper} from './form-data-wrapper';
+import {user} from './log';
+
+export class DocumentFetcher extends XhrBase {
+  /**
+   * Creates an instance of DocumentFetcher.
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super(win);
+
+    /** @private {boolean} */
+    this.viewerResponded_ = false;
+
+    /** @private {!XMLHttpRequest} */
+    this.xhr_ = new XMLHttpRequest();
+  }
+
+  /**
+   * @override
+   */
+  fetchFromNetwork_(input, init) {
+    this.viewerResponded_ = false;
+
+    return this.maybeIntercept_(input, init)
+        .then(interceptorResponse => {
+          if (interceptorResponse) {
+            this.viewerResponded_ = true;
+            return interceptorResponse;
+          }
+          // After this point, both the native `fetch` and the `fetch` polyfill
+          // will expect a native `FormData` object in the `body` property, so
+          // the native `FormData` object needs to be unwrapped.
+          if (isFormDataWrapper(init.body)) {
+            init.body = init.body.getFormData();
+          }
+
+          return (this.xhrRequest_).apply(this, arguments);
+        });
+  }
+
+  /**
+   *
+   *
+   * @param {string} input
+   * @param {!FetchInitDef} init
+   * @private
+   */
+  xhrRequest_(input, init) {
+    return new Promise((resolve, reject) => {
+      this.xhr_.open(init.method || 'GET', input, true);
+      if (init.credentials == 'include') {
+        this.xhr_.withCredentials = true;
+      }
+      this.xhr_.responseType = 'document';
+      if (init.headers) {
+        Object.keys(init.headers).forEach(header => {
+          this.xhr_.setRequestHeader(header, init.headers[header]);
+        });
+      }
+      this.xhr_.onreadystatechange = () => {
+        if (this.xhr_.readyState < /* STATUS_RECEIVED */ 2) {
+          return;
+        }
+        if (this.xhr_.status < 100 || this.xhr_.status > 599) {
+          this.xhr_.onreadystatechange = null;
+          reject(user().createExpectedError(
+              `Unknown HTTP status ${this.xhr_.status}`));
+          return;
+        }
+
+        // TODO(dvoytenko): This is currently simplified: we will wait for the
+        // whole document loading to complete. This is fine for the use cases
+        // we have now, but may need to be reimplemented later.
+        if (this.xhr_.readyState == /* COMPLETE */ 4) {
+          const options = {
+            status: this.xhr_.status,
+            statusText: this.xhr_.statusText,
+            headers: parseHeaders(this.xhr_.getAllResponseHeaders()),
+          };
+          options.url = 'responseURL' in this.xhr_
+            ? this.xhr_.responseURL : options.headers['X-Request-URL'];
+          const body = 'response' in this.xhr_
+            ? this.xhr_.response : this.xhr_.responseText;
+          resolve(new Response(/** @type {string} */ (body || ''), /** @type {!ResponseInit} */ (options)));
```

Comment:
I'm not sure this response is necessary. The only time we'd use it is in `fetchDocument` where `viewerResponded_` is true. But the only way we hit `xhrRequest_` method is if `!viewerResponded_`.

Both the header processing and the body should remain unused, and we'll only get `xhr.reponseXML`.

Reply:
Actually I agree about the body, but header processing is required in a post fetch step here: https://github.com/ampproject/amphtml/blob/e04c3bb14170a968c9f4af5e846526014a31feae/src/service/xhr-impl.js#L398-L414